### ISSUE TITLE
C stack usage [...] is too close to the limit” after running catboost multiple times

### ISSUE THREAD

**Issue by prildi** [catboost/catboost#1339]
Problem:
I have a model in RStudio witch runs different gradient boosting libraries (lgbm, xgboost, catboost). It selects a dataset, builds a model, predicts, and repeats for a different dataset. With xgb or lightgbm I have no issues, with catboost, after several model runs (several thousands) it interrupts and tells me "C stack usage [high number] is too close to the limit". The [high number] is different every time, but equivalent to several 100 GB storage. Normally, one model run consumes less than 1 GB RAM. However, after this error message, Rstudio is responding, but not do another catboost call. After restarting Rstudio, the model runs again, but just for several dozen times, then interrupts with the same error message but with another [high number]. After rebooting the machine, it runs several thousands again, next time dozens and so on. Unfortunately, it is no reproducible, sometimes it runs 10 times, sometimes 100.

For troubleshooting I tried the following, but neither worked:

-changes nvidia driver 440 to 435.
-update to the latest catboost version.
-changed all the catboost learning parameters.
-run detached() and library() for catboost, everytime I trained a model.

catboost version:0.23, 0.24
Operating System: Ubuntu 18.04 LTS
CPU: TR 1920
# GPU: RTX 2060 super

**Comment by kizill** [comment]
Could you please make a reproduction code-snippet? I've tried reproducing your issue by running train-predict functions 10k times in a loop and everything works just fine. Thank you!

**Comment by prildi** [comment]
There it is:

`library(catboost)

CAT<-function(train,test,rounds=200){
  X<-sample(nrow(train),nrow(train)/2)
  dTrain <- catboost.load_pool(data = train[-X,-ncol(train)], label = train[-X,ncol(train)])
  dTest <- catboost.load_pool(data = train[X,-ncol(train)], label = train[X,ncol(train)])
  dPred <- catboost.load_pool(test)
  PP <- list(loss_function = 'RMSE',
             od_type = "Iter",
             iterations = rounds, metric_period=round(rounds/5),
             learning_rate = 15/100,
             grow_policy = "SymmetricTree",
             logging_level = "Silent",
             task_type="GPU")
  mod <- catboost.train(dTrain,  test_pool = dTest,params = PP)
  pred<-catboost.predict(mod,dPred)
  return(pred)
}

a<-100
b<-100000
train<-matrix(runif(a*b),b,a)
test<-matrix(runif((a-1)*b),b,a-1)
for (i in 1:10000){
  erg<-CAT(train,test)
  print(i)
}`
After 49 rounds, the error message pops up:

`Error: C stack usage  237431939708 is too close to the limit`

I didn't run catboost on this system after a reboot, but I run xgb on the gpu before. This might explain the low amount of rounds after the error occurs. Using "CPU" instead of "GPU" does not lead to an error.

**Comment by kizill** [comment]
It didn't fail on my MacBook, we will try in Linux container this week. Thank you!

**Comment by prildi** [comment]
Just a little update: The error occurs even if the train matrix is very small (like 10 by 100 cells) and the error occurs during the `catboost.train` command

**Comment by kizill** [comment]
I reproduced it on linux, will fix soon.
